src/backends/cuda.jl

```diff
@@ -47,39 +47,60 @@ end
 struct CudaEvent <: Event
     event::CuEvent
 end
-
-function Event(::CUDA)
-    stream = CUDAdrv.CuDefaultStream()
-    event = CuEvent(CUDAdrv.EVENT_DISABLE_TIMING)
-    CUDAdrv.record(event, stream)
-    CudaEvent(event)
-end
-
-wait(ev::CudaEvent, progress=nothing) = wait(CPU(), ev, progress)
-function wait(::CPU, ev::CudaEvent, progress=nothing)
+function wait(ev::CudaEvent, progress=nothing)
     if progress === nothing
-        CUDAdrv.synchronize(ev.event)
+        CUDAdrv.wait(ev.event)
     else
         while !CUDAdrv.query(ev.event)
             progress()
             # do we need to `yield` here?
         end
     end
 end
+function __waitall(::CUDA, dependencies, progress, stream)
+    if dependencies isa Event
+        dependencies = (dependencies,)
+    end
+    if dependencies !== nothing
+        cudadeps  = filter(d->d isa CudaEvent,    dependencies)
+        otherdeps = filter(d->!(d isa CudaEvent), dependencies)
+        for event in cudadeps
+            CUDAdrv.wait(event.event, stream)
+        end
+        for event in otherdeps
+            wait(CUDA(), event, progress)
+        end
+    end
+end
 
-# Use this to synchronize between computation using the CuDefaultStream
-wait(::CUDA, ev::CudaEvent, progress=nothing) = __enqueue_wait(ev, CUDAdrv.CuDefaultStream())
+###
+# async_copy
+###
+function async_copy!(::CUDA, A, B; dependencies=nothing)
+    stream = next_stream()
+    __waitall(CUDA(), dependencies, yield, stream)
+    event = CuEvent(CUDAdrv.EVENT_DISABLE_TIMING)
+    GC.@preserve A B begin
+        destptr = pointer(A)
+        srcptr  = pointer(B)
+        N       = length(A)
+        unsafe_copyto!(destptr, srcptr, N, async=true, stream=stream)
+    end
+
+    CUDAdrv.record(event, stream)
 
-# There is no efficient wait for CPU->GPU synchronization, so instead we
-# do a CPU wait, and therefore block anyone from submitting more work.
-# We maybe could do a spinning wait on the GPU and atomic flag to signal from the CPU,
-# but which stream would we target?
-wait(::CUDA, ev::CPUEvent,  progress=nothing) = wait(CPU(), ev, progress)
+    return CudaEvent(event)
+end
 
-function __enqueue_wait(ev::CudaEvent, stream::CuStream)
-    CUDAdrv.wait(ev.event, stream)
+function pin!(::CUDA, a)
```

Comment:
Do the users actually need to see the `pin!(...)` function? I feel like we should just do it for the GPU-> GPU copy and not show anyone. This function is naturally inefficient on the CPU, so having the user allocate unused memory seems a bit weird.

Reply:
Yes, I think users need to see `pin!`.  Some users may want to manage their own data transfers between the host and device.  Also, I don't understand why this method is inefficient on the CPU.

Reply:
The issue is that `pin!` is costly, but we could cache the pointer...

Reply:
I just noticed there is no `pin!(::CPU, a)` function in this PR right now, so I guess if it's a GPU-specific function, that's fine. I didn't want users to allocate a new array to pin in the sake of language agnostic code.